### DISCUSSION TITLE
Which egui requirement permit virtual keyboard opening on smartphones ?

### DISCUSSION THREAD

**Discussion by buxx** [emilk/egui#1309] [Q&A] [answered]
Hello !

First, A big thank you for this amazing lib. Egui is awesome for Rust world.

I have an issue with virtual keyboard opening on smartphones when use egui with its macroquad integration. Explainations : 

Virtual keyboard opening on smartphones works well with egui when I try on my dev machine or online (tested version 0.17 and 0.16):

![Screenshot_20220226-130050](https://user-images.githubusercontent.com/1104637/155842572-ef7e3359-b907-4457-a88a-975b624f82de.png)

But, It fails to open virtual keyboard opening on smartphones when I use macroquad integration. There is the HowTo reproduction method : 

1. Init a fresh rust project `cargo init demo_smartphone_macroquad`
2. Declare these dependencies in `Cargo.toml`: 

```
[dependencies]
macroquad = "0.3"
egui = "0.16.1"
egui-macroquad = "0.8.0"
```

3. Write this `main.rs` : 

```rust
use macroquad::prelude::*;
#[macroquad::main("egui with macroquad")]
async fn main() {
    let mut text = "input here 2".to_string();
    loop {
        clear_background(WHITE);
        egui_macroquad::ui(|egui_ctx| {
            egui::Window::new("egui ❤ macroquad").show(egui_ctx, |ui| {
                ui.text_edit_singleline(&mut text);
            });
        });
        egui_macroquad::draw();
        next_frame().await;
    }
}
```

To make macroquad builds able to work with wasm-bindgem, we must use [this helper script](https://gist.github.com/olefasting/15ae263da4cf1ba308ce55c15c9b221b) : (`wget https://gist.githubusercontent.com/olefasting/15ae263da4cf1ba308ce55c15c9b221b/raw/4ca4c8e23e19ae6e0040d68b0397cb0103066736/wasm-bindgen-macroquad.sh && chmod +x wasm-bindgen-macroquad.sh`)

4. Build the wasm bin with `./wasm-bindgen-macroquad.sh demo_smartphone_macroquad --release` (wasm-bindgen-cli is required: `cargo install -f wasm-bindgen-cli`)
5. Start http server `cargo install basic-http-server && basic-http-server -a 0.0.0.0:4000 ./web`
6. Open `http://<you_computer_local_ip>:4000/` with a smartphone and try the text edit.

For information, egui macroquad integration code is visible here : https://github.com/optozorax/egui-macroquad/blob/master/src/lib.rs

**My question** : What in egui is permitting virtual keyboard opening ? Which seems to be missing in macroquad egui integration or wasm-bindgem compatibility hack. I would be happy to help make this work.

**Comment by KentaTheBugMaker** [comment, marked as answer]
egui_web attach dummy input element to body  to capture KeyboardEvent,InputEvent,CompositionEvent but macroquad does not add input element .
this workaround introduced by @n2 in #253 .

**Reply by buxx** [reply]
Thank you a lot @t18b219k for these information. I will study them about making macroquad compatible !

**Comment by n2** [comment]
The dummy input element initial/enabling/moving logic is in [this file](https://github.com/emilk/egui/blob/713917e4819d9c26068aec286ee675f7e3bff907/egui_web/src/text_agent.rs), Composition events handle logic is [here](https://github.com/emilk/egui/blob/713917e4819d9c26068aec286ee675f7e3bff907/egui/src/widgets/text_edit/builder.rs#L779), hope that helps.

**Reply by buxx** [reply]
Hello @n2 ,

I'm trying to follow the steps which "activate" this `install_text_agent` feature. There is what I understand : 

Starting from [eframe_template](https://github.com/emilk/eframe_template)

1. eframe_template indicate to wasm the `start` entrypoint. This function call `eframe::start_web`.

![2022-02-28_17-21](https://user-images.githubusercontent.com/1104637/156019210-562b8fde-fabf-47bd-9d05-c45b40179b68.png)

2. `eframe::start_web` call `egui_web::start`

![2022-02-28_17-23](https://user-images.githubusercontent.com/1104637/156019383-fb94c33d-c903-42a1-a5a0-76963918f540.png)

3. `egui_web::start` call `start_runner` which call `install_text_agent`.

![2022-02-28_17-15](https://user-images.githubusercontent.com/1104637/156019608-b400f0ad-2b70-4cb5-b8f0-a2036654fb63.png)

I suppose `#[wasm_bindgen]` on `start` in eframe_template is important too.

Am I right ?

I now have to explore [egui-macroquad](https://github.com/optozorax/egui-macroquad) and [egui-miniquad](https://github.com/not-fl3/egui-miniquad) integrations ... :p

**Comment by n2** [comment]
Yes, that's the calling process, #[wasm_bindgen] is used to expose start(...) to JavaScript.
install_text_agent(...) uses web_sys to get Document on Html and creates HtmlInputElement on it. If web_sys is available on the macroquad, you can call this install_text_agent(...) directly in the initialization function, or you have to find a proper way to create/manipulate HtmlInputElement on the macroquad.
